lib/deep_preloader.rb

```diff
@@ -3,129 +3,277 @@
+        else
+          add_association(models, association_reflection, child_spec)
+        end
+      end
+    end
+
+    def run!
+      while(@worklist.present?)
+        context, entries = @worklist.shift
+        ActiveRecord::Base.logger.debug("Preloading children in context #{context}") if DEBUG
+
+        loaded_entries, unloaded_entries = entries.partition(&:loaded?)
+
+        unloaded_keys = unloaded_entries.map(&:key).to_set.delete(nil)
+
+        ActiveRecord::Base.logger.debug("Need to load children for following keys: #{unloaded_keys.to_a}") if DEBUG
+
+        found_children = {}
+
+        if unloaded_keys.present?
+          # When :belongs_to, children could be shared with already loaded
+          # entries - use what we already have.
+          loaded_entries.each do |entry|
+            next unless entry.belongs_to?
+
+            if entry.belongs_to? && unloaded_keys.delete?(entry.key)
+              found_children[entry.key] = entry.children
+            end
+          end
+          ActiveRecord::Base.logger.debug("found loaded children for keys #{found_children.keys}") if DEBUG
+        end
+
+        if unloaded_keys.present?
+          fetched_children = context.load_children(unloaded_keys.to_a, lock: @lock)
+          ActiveRecord::Base.logger.debug("fetched children for keys #{fetched_children.keys}") if DEBUG
+          found_children.merge!(fetched_children)
+        end
+
+        unloaded_entries.each do |entry|
+          children = found_children.fetch(entry.key, [])
+          entry.children = children
+        end
+
+        entries.each do |entry|
+          children = entry.children
+          child_spec = entry.child_spec
+          next unless child_spec && children.present?
+          child_class = children.first.class # children of a given parent are all of the same type
+          add_associations_from_spec(children, child_class, child_spec)
+        end
+
+      end
+
     end
 
-    unless association_reflection.constraints.blank?
-      raise ArgumentError.new("Preloading conditional associations not supported: #{association_reflection.name}")
+    private
+
+    def add_polymorphic_association(models, association_reflection, polymorphic_child_spec)
+      assoc_name = association_reflection.name
+
+      # If a model belongs_to a polymorphic child, we know what type it is.
+      # Group models by the type of their associated child and add each
+      # separately.
+      models_by_child_class = models.group_by { |m| m.association(assoc_name).klass }
+
+      # For models with no child there's nothing to preload, but we still need
+      # to set the association target. Since we can't infer a class for
+      # `add_association`, set it up here.
+      models_by_child_class.delete(nil)&.each do |model|
+        model.association(assoc_name).loaded!
+      end
+
+      models_by_child_class.each do |child_class, child_class_models|
+        if polymorphic_child_spec
```

Comment:
But at the same time reducing the level of hinting that it's conditional. (not assign rather than assign nil). No big deal, will do anyway.